clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -5910,6 +5910,51 @@ static void handleBuiltinAliasAttr(Sema &S, Decl *D,
   D->addAttr(::new (S.Context) BuiltinAliasAttr(S.Context, AL, Ident));
 }
 
+static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  if (!AL.hasParsedType()) {
+    S.Diag(AL.getLoc(), diag::err_attribute_wrong_number_arguments) << AL << 1;
+    return;
+  }
+
+  TypeSourceInfo *ParmTSI = nullptr;
+  QualType QT = S.GetTypeFromParser(AL.getTypeArg(), &ParmTSI);
+  QT->getAsTagDecl()->setCompleteDefinitionRequired();
```

Comment:
@dwblaikie Would it be fine to do this? The intention here is that we parse type argument of the attribute, and then use it in `CGDebugInfo.h` (you might be interested in changes there as well), bypassing Sema.

Reply:
I think we want a different approach here. If we decide to do non-integral/enum types right now, I think we want to call `Sema::RequireCompleteType()` on the given type. This will then diagnose uses of types like `void` or `int[]` as well as forward declarations. Losing forward declarations is a bit sad (because it could be completed later), but I think we can relax this restriction in the future as we find uses.

Reply:
I've changed this so that `preferred_type` accepts only complete types. We can lift this up later if it's too restrictive. I just don't want to deal with consequences of incomplete types in this PR.